### PR TITLE
Add an implementation of AATree 

### PR DESCRIPTION
**Pull request by amanjpro**
I have added AATree implementation, it is basically a direct translation from the imperative algorithm shown in Wikipedia.  I've also added some useful functions like toList, inssertAll, map, filter, fold, foreach and merge.

My implementation goes for clarity not efficiency.

### CODE REVIEW

File: src/tree/AATree.scala [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+ * result in similar performance. A red-black tree is more consistent in its
+ * performance than an AA tree, but an AA tree tends to be flatter, which
+ * results in slightly faster search times
+ *
+ * Insert - O(log n)
+ * Lookup - O(log n)  
+ * Remove - O(log n)
+ */
+
+object AATree {
```
Comment (vkostyukov) [inline src/tree/AATree.scala]: You probably don't need to wrap it into an object. It's easy to read the code w/o it.

File: src/tree/AATree.scala [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+      case Fork(v, lv, l, r) if implicitly[Ordering[B]].lt(a, v) => 
+        Fork(v, lv, l.insert(a), r).rebalanceTree
+      case Fork(v, lv, l, r) if implicitly[Ordering[B]].gt(a, v) => 
+        Fork(v, lv, l, r.insert(a)).rebalanceTree
+      case _                                                     =>
+        this
+    }
+
+    def insert[B >: T: Ordering](a: B): AATree[B] = insert(a)
+    def add[B >: T: Ordering](a: B): AATree[B] = insert(a)
```
Comment (vkostyukov) [inline src/tree/AATree.scala]: I don't think we need this method. Having `insert` is enough.

File: src/tree/AATree.scala [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+        Fork(a, 1, Empty, Empty)
+      case Fork(v, lv, l, r) if implicitly[Ordering[B]].lt(a, v) => 
+        Fork(v, lv, l.insert(a), r).rebalanceTree
+      case Fork(v, lv, l, r) if implicitly[Ordering[B]].gt(a, v) => 
+        Fork(v, lv, l, r.insert(a)).rebalanceTree
+      case _                                                     =>
+        this
+    }
+
+    def insert[B >: T: Ordering](a: B): AATree[B] = insert(a)
```
Comment (vkostyukov) [inline src/tree/AATree.scala]: Why this is calling itself?

File: src/tree/AATree.scala [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+    val right: AATree[T]
+
+    def member[B >: T: Ordering](a: B): Boolean = this match {
+      case Empty                                                => false
+      case Fork(v, _, l, r) if a == v                           => true
+      case Fork(v, _, l, r) if implicitly[Ordering[B]].lt(a, v) => l.member(v)
+      case Fork(v, _, l, r)                                     => r.member(v)
+    }
+
+    def +[B >: T: Ordering](a: B): AATree[B] = this match {
```
Comment (vkostyukov) [inline src/tree/AATree.scala]: Can we just have `insert` without aliases?

File: src/tree/AATree.scala [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+      case Fork(v, lv, l, r) if implicitly[Ordering[B]].gt(a, v) => 
+        Fork(v, lv, l, r.insert(a)).rebalanceTree
+      case _                                                     =>
+        this
+    }
+
+    def insert[B >: T: Ordering](a: B): AATree[B] = insert(a)
+    def add[B >: T: Ordering](a: B): AATree[B] = insert(a)
+
+    def insertAll[B >: T: Ordering](as: List[B]): AATree[B] = {
```
Comment (vkostyukov) [inline src/tree/AATree.scala]: I think this method is redundant. We should probably focus on core functions.

File: src/tree/AATree.scala [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+      case Empty            => Empty
+      case Fork(v, _, l, r) =>
+        val mappedL = l.flatMap(f)
+        val mappedR = r.flatMap(f)
+        mappedL.merge(mappedR).merge(f(v))
+    }
+
+
+
+    def foreach(f: T => Unit): Unit = toList.foreach(f)
```
Comment (vkostyukov) [inline src/tree/AATree.scala]: These both methods `foreach` and `filter` are not really necessary here.

File: src/tree/AATree.scala [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+    def filter(p: T => Boolean): AATree[T] = Empty.insertAll(toList.filter(p))
+     
+
+    def fold[B: Ordering](f: (T, B, B) => B, z: B): B = this match {
+      case Empty            => z
+      case Fork(v, _, l, r) =>
+        f(v, l.fold(f, z), r.fold(f, z))
+    }
+
+    def delete[B >: T: Ordering](a: B): AATree[B] = this match {
```
Comment (vkostyukov) [inline src/tree/AATree.scala]: It should probably be named `remove` instead. `insert`/`remove` but `create`/`delete`.

File: src/tree/AATree.scala [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+        val succ = successor
+        val r2 = r.delete(succ)
+        Fork(succ, lv, Empty, r2).rebalanceAfterDelete
+      case Fork(_, lv, l, r)                                     =>
+        val pred = predecessor
+        val l2 = l.delete(pred)
+        Fork(pred, lv, l2, r).rebalanceAfterDelete
+    }
+
+    def -[B >: T: Ordering](a: B): AATree[B] = delete(a)
```
Comment (vkostyukov) [inline src/tree/AATree.scala]: This is not necessary.

**Comment by vkostyukov** [comment]
Thanks a lot @amanjpro! This looks fantastic! Just a couple of nits form my side. Also, please, don't forget to add a new tree implementation in to README.

File: src/tree/AATree.scala (hunk 1 above)
Comment (amanjpro) [inline src/tree/AATree.scala]: Sure, I did that to add an implicit evidence for Ordering[Nothing]. But I added it to its companion object.

File: src/tree/AATree.scala (hunk 4 above)
Comment (amanjpro) [inline src/tree/AATree.scala]: Done

File: src/tree/AATree.scala (hunk 3 above)
Comment (amanjpro) [inline src/tree/AATree.scala]: Opps, my bad

File: src/tree/AATree.scala (hunk 2 above)
Comment (amanjpro) [inline src/tree/AATree.scala]: Got rid of the aliases

File: src/tree/AATree.scala (hunk 5 above)
Comment (amanjpro) [inline src/tree/AATree.scala]: I needed this to do merge, but I changed my merge function now not to rely on insert all

File: src/tree/AATree.scala (hunk 6 above)
Comment (amanjpro) [inline src/tree/AATree.scala]: They are gone now

File: src/tree/AATree.scala (hunk 7 above)
Comment (amanjpro) [inline src/tree/AATree.scala]: The functions in Wikipedia are called insert/delete. I didn't want to change it

File: src/tree/AATree.scala (hunk 8 above)
Comment (amanjpro) [inline src/tree/AATree.scala]: Gone

**Comment by amanjpro** [comment]
Done!

File: src/tree/AATree.scala [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+    }
+    val t4 = t3.split
+    t4 match {
+      case Fork(v, lv, l, r) => Fork(v, lv, l, r.split)
+      case _                 => t4
+    }
+  }
+
+  protected def successor[B >: T: Ordering]: B = {
+    def successor2(t: AATree[B]): B = this match {
```
Comment (vkostyukov) [inline src/tree/AATree.scala]: This is usually called `loop`.

File: src/tree/AATree.scala [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+    this match {
+      case Fork(_, _, _, Fork(v, _, Empty, _)) => v
+      case Fork(_, _, _, Fork(v, _, l, _))     => successor2(l)
+      case _                                   =>
+        throw new Exception("Empty node does not have a successor")
+    }
+  }
+
+  protected def predecessor[B >: T: Ordering]: B = {
+    def predecessor2(t: AATree[B]): B = this match {
```
Comment (vkostyukov) [inline src/tree/AATree.scala]: We can call it `loop`.

File: src/tree/AATree.scala (hunk 9 above)
Comment (amanjpro) [inline src/tree/AATree.scala]: Do you mean the inner function?

**Comment by vkostyukov** [comment]
Just a couple of tiny things. @amanjpro could you also please add comments to each method with time/memory requirements.

```scala
/**
 * Time - O(log n)
 * Memory - O(log n) 
 */
def insert()
```

File: src/tree/AATree.scala (hunk 9 above)
Comment (vkostyukov) [inline src/tree/AATree.scala]: Yep. It's not visible outside `successor`, so we can use unified name for these functions. It usually `loop`.

**Comment by amanjpro** [comment]
@vkostyukov is it better now?

**Comment by vkostyukov** [comment]
Yes. :+1: Merging this.

**Pull request merged**